# Lab Customer Analysis Round 5 & 6

## **Processing Data**

1. One Hot/Label Encoding (categorical).
2. Concat DataFrames
3. X-y split. 
4. Feature scaling - Normalize (numerical). 
5. Train-test split.
6. Apply linear regression.
7. Model Validation
    - R2.
    - MSE.
    - RMSE.
    - MAE.

In [408]:
import numpy as np
import pandas as pd
import seaborn as sns
from scipy import stats
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, Normalizer, StandardScaler
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
import statsmodels.api as sm
from statsmodels.formula.api import ols


In [409]:
df = pd.read_csv('/Users/dooinnkim/ironhack_da_may_2023/lab-customer-analysis-round-5/files_for_lab/csv_files/marketing_customer_analysis.csv')

In [410]:
df.head()

Customer       State  Customer Lifetime Value Response  Coverage Education  \
0  BU79786  Washington              2763.519279       No     Basic  Bachelor   
1  QZ44356     Arizona              6979.535903       No  Extended  Bachelor   
2  AI49188      Nevada             12887.431650       No   Premium  Bachelor   
3  WW63253  California              7645.861827       No     Basic  Bachelor   
4  HB64268  Washington              2813.692575       No     Basic  Bachelor   

  Effective To Date EmploymentStatus Gender  Income  ...  \
0           2/24/11         Employed      F   56274  ...   
1           1/31/11       Unemployed      F       0  ...   
2           2/19/11         Employed      F   48767  ...   
3           1/20/11       Unemployed      M       0  ...   
4            2/3/11         Employed      M   43836  ...   

  Months Since Policy Inception Number of Open Complaints  Number of Policies  \
0                             5                         0                   1   
1                            42                         0                   8   
2                            38                         0                   2   
3                            65                         0                   7   
4                            44                         0                   1   

      Policy Type        Policy  Renew Offer Type  Sales Channel  \
0  Corporate Auto  Corporate L3            Offer1          Agent   
1   Personal Auto   Personal L3            Offer3          Agent   
2   Personal Auto   Personal L3            Offer1          Agent   
3  Corporate Auto  Corporate L2            Offer1    Call Center   
4   Personal Auto   Personal L1            Offer1          Agent   

  Total Claim Amount  Vehicle Class Vehicle Size  
0         384.811147   Two-Door Car      Medsize  
1        1131.464935  Four-Door Car      Medsize  
2         566.472247   Two-Door Car      Medsize  
3         529.881344            SUV      Medsize  
4         138.130879  Four-Door Car      Medsize  

[5 rows x 24 columns]

In [411]:
df.shape

(9134, 24)

In [412]:
df.isna().sum()

Customer                         0
State                            0
Customer Lifetime Value          0
Response                         0
Coverage                         0
Education                        0
Effective To Date                0
EmploymentStatus                 0
Gender                           0
Income                           0
Location Code                    0
Marital Status                   0
Monthly Premium Auto             0
Months Since Last Claim          0
Months Since Policy Inception    0
Number of Open Complaints        0
Number of Policies               0
Policy Type                      0
Policy                           0
Renew Offer Type                 0
Sales Channel                    0
Total Claim Amount               0
Vehicle Class                    0
Vehicle Size                     0
dtype: int64

In [413]:
df.columns = [column.lower().replace(' ', '_') for column in df.columns]

In [414]:
df.dtypes

customer                          object
state                             object
customer_lifetime_value          float64
response                          object
coverage                          object
education                         object
effective_to_date                 object
employmentstatus                  object
gender                            object
income                             int64
location_code                     object
marital_status                    object
monthly_premium_auto               int64
months_since_last_claim            int64
months_since_policy_inception      int64
number_of_open_complaints          int64
number_of_policies                 int64
policy_type                       object
policy                            object
renew_offer_type                  object
sales_channel                     object
total_claim_amount               float64
vehicle_class                     object
vehicle_size                      object
dtype: object

In [415]:
df['effective_to_date'] = pd.to_datetime(df['effective_to_date'])

In [416]:
df.dtypes

customer                                 object
state                                    object
customer_lifetime_value                 float64
response                                 object
coverage                                 object
education                                object
effective_to_date                datetime64[ns]
employmentstatus                         object
gender                                   object
income                                    int64
location_code                            object
marital_status                           object
monthly_premium_auto                      int64
months_since_last_claim                   int64
months_since_policy_inception             int64
number_of_open_complaints                 int64
number_of_policies                        int64
policy_type                              object
policy                                   object
renew_offer_type                         object
sales_channel                           

## 1. One Hot/Label Encoding (categorical).

In [417]:
df_oneHotCoding = df.copy()
df_dummy = df.copy()



In [420]:
list(df_dummy.columns)
 

['customer',
 'state',
 'customer_lifetime_value',
 'response',
 'coverage',
 'education',
 'effective_to_date',
 'employmentstatus',
 'gender',
 'income',
 'location_code',
 'marital_status',
 'monthly_premium_auto',
 'months_since_last_claim',
 'months_since_policy_inception',
 'number_of_open_complaints',
 'number_of_policies',
 'policy_type',
 'policy',
 'renew_offer_type',
 'sales_channel',
 'total_claim_amount',
 'vehicle_class',
 'vehicle_size']

In [421]:
list(df_oneHotCoding.columns)

['customer',
 'state',
 'customer_lifetime_value',
 'response',
 'coverage',
 'education',
 'effective_to_date',
 'employmentstatus',
 'gender',
 'income',
 'location_code',
 'marital_status',
 'monthly_premium_auto',
 'months_since_last_claim',
 'months_since_policy_inception',
 'number_of_open_complaints',
 'number_of_policies',
 'policy_type',
 'policy',
 'renew_offer_type',
 'sales_channel',
 'total_claim_amount',
 'vehicle_class',
 'vehicle_size']

In [422]:
# exclude customer and effective_to_date
model_sample1 = ['state',
 'customer_lifetime_value',
 'response',
 'coverage',
 'education',
 'employmentstatus',
 'gender',
 'income',
 'location_code',
 'marital_status',
 'monthly_premium_auto',
 'months_since_last_claim',
 'months_since_policy_inception',
 'number_of_open_complaints',
 'number_of_policies',
 'policy_type',
 'policy',
 'renew_offer_type',
 'sales_channel',
 'total_claim_amount',
 'vehicle_class',
 'vehicle_size']

# multicollinearlity and negative correlated features removed ('number_of_policies', 'number_of_open_complaints', 'income')
model_sample2 = ['state',
 'customer_lifetime_value',
 'response',
 'coverage',
 'education',
 'employmentstatus',
 'gender',
 'location_code',
 'marital_status',
 'monthly_premium_auto',
 'months_since_last_claim',
 'months_since_policy_inception',
 'policy_type',
 'policy',
 'renew_offer_type',
 'sales_channel',
 'total_claim_amount',
 'vehicle_class',
 'vehicle_size']

df_oneHotCoding=df_oneHotCoding[model_sample2]
df_dummy=df_dummy[model_sample2]



In [423]:
df_oneHotCoding.head()

state  customer_lifetime_value response  coverage education  \
0  Washington              2763.519279       No     Basic  Bachelor   
1     Arizona              6979.535903       No  Extended  Bachelor   
2      Nevada             12887.431650       No   Premium  Bachelor   
3  California              7645.861827       No     Basic  Bachelor   
4  Washington              2813.692575       No     Basic  Bachelor   

  employmentstatus gender location_code marital_status  monthly_premium_auto  \
0         Employed      F      Suburban        Married                    69   
1       Unemployed      F      Suburban         Single                    94   
2         Employed      F      Suburban        Married                   108   
3       Unemployed      M      Suburban        Married                   106   
4         Employed      M         Rural         Single                    73   

   months_since_last_claim  months_since_policy_inception     policy_type  \
0                       32                              5  Corporate Auto   
1                       13                             42   Personal Auto   
2                       18                             38   Personal Auto   
3                       18                             65  Corporate Auto   
4                       12                             44   Personal Auto   

         policy renew_offer_type sales_channel  total_claim_amount  \
0  Corporate L3           Offer1         Agent          384.811147   
1   Personal L3           Offer3         Agent         1131.464935   
2   Personal L3           Offer1         Agent          566.472247   
3  Corporate L2           Offer1   Call Center          529.881344   
4   Personal L1           Offer1         Agent          138.130879   

   vehicle_class vehicle_size  
0   Two-Door Car      Medsize  
1  Four-Door Car      Medsize  
2   Two-Door Car      Medsize  
3            SUV      Medsize  
4  Four-Door Car      Medsize

In [424]:
list(df_oneHotCoding.select_dtypes("object"))

['state',
 'response',
 'coverage',
 'education',
 'employmentstatus',
 'gender',
 'location_code',
 'marital_status',
 'policy_type',
 'policy',
 'renew_offer_type',
 'sales_channel',
 'vehicle_class',
 'vehicle_size']

In [425]:
encoder = OneHotEncoder(drop='first')

categorical_var = list(df_oneHotCoding.select_dtypes("object"))

encoder.fit(df_oneHotCoding[categorical_var])

cols = []
for i in range(len(categorical_var)):
    cols += list(encoder.categories_[i][1:])


df_encoded = pd.DataFrame(encoder.transform(df_oneHotCoding[categorical_var]).todense(), columns=cols, index=df_oneHotCoding.index)




In [426]:
# if using get_dummies for Encoding

df_dummy = pd.get_dummies(df_dummy, drop_first=True)

In [427]:
df_dummy.head()

customer_lifetime_value  monthly_premium_auto  months_since_last_claim  \
0              2763.519279                    69                       32   
1              6979.535903                    94                       13   
2             12887.431650                   108                       18   
3              7645.861827                   106                       18   
4              2813.692575                    73                       12   

   months_since_policy_inception  total_claim_amount  state_California  \
0                              5          384.811147                 0   
1                             42         1131.464935                 0   
2                             38          566.472247                 0   
3                             65          529.881344                 1   
4                             44          138.130879                 0   

   state_Nevada  state_Oregon  state_Washington  response_Yes  ...  \
0             0             0                 1             0  ...   
1             0             0                 0             0  ...   
2             1             0                 0             0  ...   
3             0             0                 0             0  ...   
4             0             0                 1             0  ...   

   sales_channel_Branch  sales_channel_Call Center  sales_channel_Web  \
0                     0                          0                  0   
1                     0                          0                  0   
2                     0                          0                  0   
3                     0                          1                  0   
4                     0                          0                  0   

   vehicle_class_Luxury Car  vehicle_class_Luxury SUV  vehicle_class_SUV  \
0                         0                         0                  0   
1                         0                         0                  0   
2                         0                         0                  0   
3                         0                         0                  1   
4                         0                         0                  0   

   vehicle_class_Sports Car  vehicle_class_Two-Door Car  vehicle_size_Medsize  \
0                         0                           1                     1   
1                         0                           0                     1   
2                         0                           1                     1   
3                         0                           0                     1   
4                         0                           0                     1   

   vehicle_size_Small  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  

[5 rows x 48 columns]

## 2. Concat DataFrames

In [428]:
df_oneHotCoding.drop(categorical_var, axis=1, inplace=True)

df_oneHotCoding = pd.concat([df_oneHotCoding, df_encoded], axis=1)

In [429]:
df_oneHotCoding.head()

customer_lifetime_value  monthly_premium_auto  months_since_last_claim  \
0              2763.519279                    69                       32   
1              6979.535903                    94                       13   
2             12887.431650                   108                       18   
3              7645.861827                   106                       18   
4              2813.692575                    73                       12   

   months_since_policy_inception  total_claim_amount  California  Nevada  \
0                              5          384.811147         0.0     0.0   
1                             42         1131.464935         0.0     0.0   
2                             38          566.472247         0.0     1.0   
3                             65          529.881344         1.0     0.0   
4                             44          138.130879         0.0     0.0   

   Oregon  Washington  Yes  ...  Branch  Call Center  Web  Luxury Car  \
0     0.0         1.0  0.0  ...     0.0          0.0  0.0         0.0   
1     0.0         0.0  0.0  ...     0.0          0.0  0.0         0.0   
2     0.0         0.0  0.0  ...     0.0          0.0  0.0         0.0   
3     0.0         0.0  0.0  ...     0.0          1.0  0.0         0.0   
4     0.0         1.0  0.0  ...     0.0          0.0  0.0         0.0   

   Luxury SUV  SUV  Sports Car  Two-Door Car  Medsize  Small  
0         0.0  0.0         0.0           1.0      1.0    0.0  
1         0.0  0.0         0.0           0.0      1.0    0.0  
2         0.0  0.0         0.0           1.0      1.0    0.0  
3         0.0  1.0         0.0           0.0      1.0    0.0  
4         0.0  0.0         0.0           0.0      1.0    0.0  

[5 rows x 48 columns]

## 3. X-y split. 

In [430]:
X = df_dummy.drop(['total_claim_amount'], axis=1)
y = df_dummy['total_claim_amount']

# X = df_oneHotCoding.drop(['total_claim_amount'], axis=1)
# y = df_oneHotCoding['total_claim_amount']

In [431]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [432]:
X_train

customer_lifetime_value  monthly_premium_auto  months_since_last_claim  \
434               5015.009472                   130                       12   
4641              5149.301306                   131                        5   
4952              4904.894731                   139                        3   
1489              8510.525936                   121                        5   
812               3278.531880                    83                       13   
...                       ...                   ...                      ...   
5734              7334.328083                    61                       31   
5191              5498.940679                    73                       17   
5390              8992.779137                   129                       13   
860              14635.451580                   139                        5   
7270              3349.639274                    84                       34   

      months_since_policy_inception  state_California  state_Nevada  \
434                              15                 0             0   
4641                              2                 0             0   
4952                             51                 1             0   
1489                             94                 0             0   
812                              19                 0             0   
...                             ...               ...           ...   
5734                             63                 1             0   
5191                             64                 1             0   
5390                              4                 1             0   
860                              56                 0             0   
7270                             13                 0             0   

      state_Oregon  state_Washington  response_Yes  coverage_Extended  ...  \
434              0                 1             1                  1  ...   
4641             1                 0             0                  1  ...   
4952             0                 0             1                  1  ...   
1489             0                 0             0                  1  ...   
812              0                 0             0                  1  ...   
...            ...               ...           ...                ...  ...   
5734             0                 0             0                  0  ...   
5191             0                 0             1                  0  ...   
5390             0                 0             0                  1  ...   
860              0                 0             1                  1  ...   
7270             0                 0             0                  1  ...   

      sales_channel_Branch  sales_channel_Call Center  sales_channel_Web  \
434                      0                          0                  0   
4641                     0                          0                  0   
4952                     0                          0                  0   
1489                     1                          0                  0   
812                      0                          1                  0   
...                    ...                        ...                ...   
5734                     0                          0                  0   
5191                     1                          0                  0   
5390                     0                          1                  0   
860                      0                          1                  0   
7270                     1                          0                  0   

      vehicle_class_Luxury Car  vehicle_class_Luxury SUV  vehicle_class_SUV  \
434                          0                         0                  1   
4641                         0                         0                  1   
4952                         0                         0                  0   
1489                      

## 4. Feature scaling - Normalize (numerical)

In [433]:
scaler = StandardScaler()
scaler.fit(X_train)

StandardScaler()

## 5. Train-test split.

In [436]:
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)


## 6. Apply linear regression.

In [435]:
# Start the model object:
lm = LinearRegression()

# Fit the model object on the training set:
lm.fit(X_train, y_train)

LinearRegression()

## 7. Model Validation

In [437]:
# Get predictions for the test set:
predictions = lm.predict(X_test)
# Calculate your metrics:
rmse = mean_squared_error(y_test, predictions, squared=False) # or mse with squared=True
mae = mean_absolute_error(y_test, predictions)
print("R2_score:", round(r2_score(y_test, predictions),2)) 
print("RMSE:", rmse)
print("MAE:", mae)

R2_score: 0.76
RMSE: 138.57322038064123
MAE: 94.5838190635902


# Test Note

#### Which one is better model??? > Have no idea how to interprete this...



### 1. Model Sample1: 

- R2_score: 0.77
- RMSE: 138.50093390700894
- MAE: 94.52272816801197

### 2. Model Sample 2: removed multicoliniarlity and negative correlation in numerical features

- R2_score: 0.76
- RMSE: 138.57322038064123
- MAE: 94.5838190635902



